File: correct-crypto/typescript/aeadEncrypt.ts
// Correct AES-256-GCM authenticated encryption with WebCrypto (SubtleCrypto).
export async function generateKey(): Promise<CryptoKey> {
  return crypto.subtle.generateKey({ name: 'AES-GCM', length: 256 }, true, ['encrypt', 'decrypt']);
}

export async function encrypt(key: CryptoKey, plaintext: Uint8Array, aad: Uint8Array): Promise<Uint8Array> {
  const iv = crypto.getRandomValues(new Uint8Array(12));   // unique 96-bit nonce
  const ct = await crypto.subtle.encrypt(
    { name: 'AES-GCM', iv, additionalData: aad, tagLength: 128 },
    key,
    plaintext,
  );
  const out = new Uint8Array(iv.length + ct.byteLength);
  out.set(iv, 0);
  out.set(new Uint8Array(ct), iv.length);
  return out;
}

File: correct-crypto/rust/aead_encrypt.rs
// Correct AES-256-GCM authenticated encryption with a random nonce (aes-gcm crate).
use aes_gcm::aead::{Aead, AeadCore, KeyInit, OsRng};
use aes_gcm::{Aes256Gcm, Nonce};

pub fn encrypt(key_bytes: &[u8; 32], plaintext: &[u8]) -> Vec<u8> {
    let cipher = Aes256Gcm::new(key_bytes.into());
    let nonce = Aes256Gcm::generate_nonce(&mut OsRng);  // fresh random 96-bit nonce
    let mut out = nonce.to_vec();
    let ciphertext = cipher.encrypt(&nonce, plaintext).expect("encryption failure");
    out.extend_from_slice(&ciphertext);
    out
}

pub fn decrypt(key_bytes: &[u8; 32], blob: &[u8]) -> Vec<u8> {
    let (nonce, ciphertext) = blob.split_at(12);
    let cipher = Aes256Gcm::new(key_bytes.into());
    cipher.decrypt(Nonce::from_slice(nonce), ciphertext).expect("decryption failure")
}
